content/en/docs/advanced-tasks/fail-forward-upgrades.md

```diff
@@ -0,0 +1,88 @@
+- If a CSV is claimed by a Subscription, the resolver will allow it to be upgraded by a bundle that replaces it.
+- If a CSV is not claimed by a Subscription, it is assumed that the user installed the CSV themselves, that OLM should not upgrade the CSV, and that it must appear in the next set of installables.
+
+OLM's Resolver is deterministic, meaning that the set of installables will not change for a given set of arguments if no new upgrades have been declared in the existing CatalogSources. When an upgrade fails due to a CSV entering the FAILED phase, the CSV being replaced still exists and is not referenced by a subscription. Today, OLM would send both `Operator v1` and `Operator v2` to the resolver which will normally be unsatisfiable because `Operator v1` is marked as required and `Operator v2` cannot be upgraded further since it provides the same APIs as `Operator v1`. In order to support "failing forward", OLM needs to omit CSVs in the REPLACING phase from the set of arguments sent to the resolver when "fail forward" upgrades are enabled.
+
+### When an InstallPlan failed
+
+Let's review what steps must be taken to recover from a failed InstallPlan today:
+1. `Operator v1` is being upgraded to `Operator v2`.
+2. The InstallPlan for `Operator v2` fails.
+3. `Operator v3` is added to the catalog and is defined as the upgrade for `Operator v1`.
+4. The user deletes the InstallPlan created for `Operator v2`.
+5. A new InstallPlan is generated for `Operator v3` and the upgrade succeeds.
+
+When you opt into "fail forward" upgrades, OLM allows new InstallPlans to be generated if:
+- The failed installPlan is referenced by a subscription in the namespace.
+- The catalogSource has been updated to include a new upgrade for the set of arguments.
+
+In practice, the fourth step from the previous workflow would be removed, meaning that the you simply need to update the CatalogSource and wait for the cluster to move past the failed install.
+
+>BEST PRACTICE NOTE: If a bundle is known to fail, it should be skipped in the upgrade graph using the [skips](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#skips) or [skipRange](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#skiprange) feature.
+
+### When one or more CSVs have failed
+
+Let's review what steps must be taken to recover from a failed CSV today:
+1. `Operator v1` is being upgraded to `Operator v2`.
+2. The CSV for `Operator v2` enters the FAILED phase.
+3. `Operator v3` is added to the catalog which replaces or skips `Operator v2`.
+3. The resolver cannot upgrade `Operator v2` while including `Operator v1` in the solution set, upgrade is blocked.
+4. User manually deletes the existing CSVs, a new InstallPlan is generated and approved.
+5. `Operator v3` is installed and the upgrade succeeds.
```

Comment:
Would this benefit from a mermaid node graph?

Reply:
It's funny I said the same thing on the EP lol. Yes, it definitely would, I'll get on adding that.

Reply:
https://learn.netlify.app/en/shortcodes/mermaid/ ?

Reply:
Alright, I added some mermaid graphs. Thoughts?